crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+//! # Synapse Database Reader
+//!
+//! This module provides facilities for streaming relevant types of database records from a Synapse database.
+
+use async_stream::stream;
+use chrono::{DateTime, Utc};
+use futures_util::Stream;
+use sea_query::{enum_def, Expr, Iden, PostgresQueryBuilder, Query};
+use sea_query_binder::SqlxBinder;
+use sqlx::{query, query_with, FromRow, PgConnection, Postgres, Row, Type};
+use thiserror::Error;
+use thiserror_ext::ContextInto;
+
+#[derive(Debug, Error, ContextInto)]
+pub enum Error {
+    #[error("database error whilst {context}: {source}")]
+    Database {
+        source: sqlx::Error,
+        context: String,
+    },
+}
+
+#[derive(Clone, Debug, sqlx::Decode)]
+pub struct FullUserId(pub String);
+
+impl Type<Postgres> for FullUserId {
+    fn type_info() -> <sqlx::Postgres as sqlx::Database>::TypeInfo {
+        <String as Type<Postgres>>::type_info()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ExtractLocalpartError {
+    #[error("user ID does not start with `@` sigil")]
+    NoAtSigil,
+    #[error("user ID does not have a `:` separator")]
+    NoSeparator,
+    #[error("wrong server name: expected {expected:?}, got {found:?}")]
+    WrongServerName { expected: String, found: String },
+}
+
+impl FullUserId {
+    /// Extract the localpart from the User ID, asserting that the User ID has the correct
+    /// server name.
+    ///
+    /// # Errors
+    ///
+    /// A handful of basic validity checks are performed and an error may be returned
+    /// if the User ID is not valid.
+    /// However, the User ID grammar is not checked fully.
+    ///
+    /// If the wrong server name is asserted, returns an error.
+    pub fn extract_localpart(
+        &self,
+        expected_server_name: &str,
+    ) -> Result<&str, ExtractLocalpartError> {
+        let Some(without_sigil) = self.0.strip_prefix('@') else {
+            return Err(ExtractLocalpartError::NoAtSigil);
+        };
+
+        let Some((localpart, server_name)) = without_sigil.split_once(':') else {
+            return Err(ExtractLocalpartError::NoSeparator);
+        };
+
+        if server_name != expected_server_name {
+            return Err(ExtractLocalpartError::WrongServerName {
+                expected: expected_server_name.to_owned(),
+                found: server_name.to_owned(),
+            });
+        };
+
+        Ok(localpart)
+    }
+}
+
+/// A Synapse boolean.
+/// Synapse stores booleans as 0 or 1, due to compatibility with old SQLite versions
+/// that did not have native boolean support.
+#[derive(Clone, Debug)]
+pub struct SynapseBool(pub bool);
```

Comment:
I think I would have gone for a private field + `impl From<SynapseBool> for bool`, mainly because I don't like having to access tuples through `.0`, but that's fiiiine